CMakeLists.txt

```diff
@@ -91,9 +91,10 @@ elseif(NOT LIBHDBPP_PROJECT_BUILD AND NOT FETCH_LIBHDBPP)
     find_package(${BACKEND_NAME})
 
     # The find package must set the backend target variable to be valid
-    if(NOT ${BACKEND_TARGET})
-	    message(STATUS "ERROR ${PACKAGES_FOUND}")
-    endif(NOT ${BACKEND_TARGET})
+    if(NOT DEFINED BACKEND_TARGET)
```

Comment:
Why BACKEND_TARGET is mandatory? Cannot hdbpp-es be compiled anymore independently from the backend?

Reply:
Nope. This is one of the change I'd like to push, we link directly against a backend. libhdbpp is only the header AbstractDb.h, so hdbpp-es now link on a real implementation.

Reply:
OK, but which are the pros? Cons are for instance that to test on the same machine different backends we need to have different binaries.

Reply:
Just to chip in. 

Its a simple architecture that works for 99% of the people 99% of the time. There is no need for a complex architecture just to allow slightly easier testing. Future users will thank you if they can setup and run quickly.

Reply:
And this way you can detect some issues at build time instead of runtime, which is better imho.
Actually if we want to keep the ability to load different backends from the same binary I would change libhdbpp so that the build system detects installed backends and manage the switch somehow.

Reply:
Well, it's not just matter of testing. The idea behind loading the back-end at runtime is that you can run the same executable, on the same platform, at the same time and use different back-ends for different purposes. If we link at compile time... we end up with different executable. The if I want to change the back-end I have to deploy another device. I'd prefer not to.

Reply:
As I understand it, the event subscriber needs to be restarted if you change the back end anyway. Unless this has changed recently? 

Is using multiple back ends of a running executable a common enough scenario that you should introduce this level of additional complexity to it? Do you have users running multiple database clusters to support this type of operation?

Reply:
I think the more likely scenario of a "complex setup" could be different hdbpp-es instances running on the same machine and pointing to different backends, at least this is what I use for benchmarking and that could be not so unusual. No evidence of anyone else doing that in any case.

Reply:
Ok, so, this is what Docker can address.

I am not on this project anymore (or at the ESRF), so really you should not listen to me :) 

I will say, as advice, consider you users and use cases. I have awful memories of trying to get this lot up and running together. This is why I set up that combined project. 

There is no reason why you cant support both. I had a branch doing just that at one point, no idea what happened to it. This is why libhdbpp-timescale includes the libhdbpp headers.

Reply:
We can have one of the backend being libhdbpp itself, allowing for dynamic library loading.

But if what your need is to have an executable that can load different backends, I would rather have it linked to all the backends installed, and not rely on dynamic loading. This can be done at the build level.

Reply:
It's an option, but implies that you have all the backends always built and installe for all the platforma/architectures. Moreover, implies that you have all the thirdly party libraries installed when you build. This is something I'd prefer to avoid. For instance, it will prevent to build an event subscriber for our PPC platforms (where there is no support for cassandra or timescale).

Reply:
Agreed, if you want to add a new backend, you'll have to recompile your executable.

Apart from this, it is just a matter of moving the moment you see you have a missing library to the building time rather than the runtime, which I find more interesting.

Anyway, I'll add libhdbpp as a possible backend to have dynamic loading capabilities, we can do both.

Reply:
I updated libhdbpp and this project to allow dynamic loading. You can try by fetching libhdbpp, tag origin/project-build

Reply:
So actually
`cmake -DFETCH_LIBHDBPP=ON -DLIBHDBPP_BACKEND=libhdbpp -DFETCH_LIBHDBPP_TAG=project-build ..`
works to compile against libhdbpp and loading dynamically backend implementation.
But how to have the same result with -DFETCH_LIBHDBPP=OFF ?
`cmake -DLIBHDBPP_BACKEND=libhdbpp -DCMAKE_INCLUDE_PATH=some/libhdbpp/include/path -DCMAKE_LIBRARY_PATH=some/libhdbpp/lib/path ..`
complains about "Backend libhdbpp does not define the BACKEND_TARGET variable needed to build."

Some more thoughts about project-build.
HDB++ is not just timescaledb, cassandra, mysql: in this same tango-controls-hdbpp there are also libhdbpp-mysql-legacy, libhdbpp-postgresql, libhdbpp-elk and possibly others outside.
It was designed this way also to let people as free as possible to create and use their own backend implementations.
I think also that its default should remain that cmake is not mandatory for backend implementations.

Reply:
Indeed I wasn't exporting the target for the shared library. It should be fixed in the libhdbpp project-build branch.

Reply:
OK, so
`cmake -DLIBHDBPP_BACKEND=libhdbpp -DCMAKE_PREFIX_PATH=/some/libhdbpp/install/path ..`
works.
Compiling with FETCH_LIBHDBPP=OFF (which is the default) and linking to some backed implementation (mysql for instance) needs the same modifications done on libhdbpp ?

Reply:
I think you already have them. libhdbpp was a bit special as I set it up so that it doesn't compile anything, the library was not exported on installation, I forgot about it.